cirq/ops/gate_operation.py

```diff
@@ -52,8 +55,8 @@ def qubits(self) -> Tuple['cirq.Qid', ...]:
         """The qubits targeted by the operation."""
         return self._qubits
 
-    def with_qubits(self, *new_qubits: 'cirq.Qid') -> 'cirq.Operation':
-        return self.gate.on(*new_qubits)
+    def with_qubits(self: TSelf, *new_qubits: 'cirq.Qid') -> TSelf:
+        return cast(TSelf, self.gate.on(*new_qubits))
```

Comment:
Optional: Have you tried making `GateOperation` parametrized with the type of gate it wraps? I think this should let us remove the cast.

Reply:
I didn't try that, but my experience with "genericizing" classes in python is that it is not trivial, especially for something this widely used. The relationship between gates and operations is also complicated by various specialized operation subclasses that we have.